src/System.Private.CoreLib/shared/System/Runtime/InteropServices/SafeBuffer.cs

```diff
@@ -110,13 +110,20 @@ public void Initialize(ulong numBytes)
         [CLSCompliant(false)]
         public void Initialize(uint numElements, uint sizeOfEachElement)
         {
-            if (IntPtr.Size == 4 && numElements * sizeOfEachElement > uint.MaxValue)
-                throw new ArgumentOutOfRangeException("numBytes", SR.ArgumentOutOfRange_AddressSpace);
+            if (sizeOfEachElement == 0)
+            {
+                _numBytes = (UIntPtr)0;
+            }
+            else
+            {
+                if (IntPtr.Size == 4 && numElements > uint.MaxValue / sizeOfEachElement)
+                    throw new ArgumentOutOfRangeException("numBytes", SR.ArgumentOutOfRange_AddressSpace);
 
-            if (numElements * sizeOfEachElement >= (ulong)Uninitialized)
-                throw new ArgumentOutOfRangeException(nameof(numElements), SR.ArgumentOutOfRange_UIntPtrMax);
+                if (numElements >= (ulong)Uninitialized / sizeOfEachElement)
+                    throw new ArgumentOutOfRangeException(nameof(numElements), SR.ArgumentOutOfRange_UIntPtrMax);
 
-            _numBytes = checked((UIntPtr)(numElements * sizeOfEachElement));
+                _numBytes = checked((UIntPtr)(numElements * sizeOfEachElement)); 
```

Comment:
We would have detected the overflow problem here anyway. So the net effect of the fix is that we will throw ArgumentOutOfRangeException instead of OverflowException. Is it correct?

Reply:
Actually, there is another bug here on 64-bit platforms. If `numElements=0x4000_0000` and `sizeOfEachElement=0x100`, we are going to throw exception for no good reason.